game/gothic.cpp

```diff
@@ -751,176 +735,141 @@ std::u16string Gothic::nestedPath(const std::initializer_list<const char16_t*> &
+  vm.register_external("printdialog",      [this](int dialognr, std::string_view msg, int posx, int posy, std::string_view font, int timesec){ return printdialog(dialognr, msg, posx, posy, font, timesec); });
+  vm.register_external("print",            [this](std::string_view msg){ print(msg); });
 
-  vm.registerExternalFunction("exitgame",            [this](Daedalus::DaedalusVM& vm){ exitgame(vm);             });
+  vm.register_external("doc_create",          [this](){ return doc_create(); });
+  vm.register_external("doc_createmap",       [this](){ return doc_createmap(); });
+  vm.register_external("doc_setpage",         [this](int handle, int page, std::string_view img, int scale){ doc_setpage(handle, page, img, scale); });
+  vm.register_external("doc_setpages",        [this](int handle, int count){ doc_setpages(handle, count); });
+  vm.register_external("doc_printline",       [this](int handle, int page, std::string_view text){ doc_printline(handle, page, text); });
+  vm.register_external("doc_printlines",      [this](int handle, int page, std::string_view text){ doc_printlines(handle, page, text); });
+  vm.register_external("doc_setmargins",      [this](int handle, int page, int left, int top, int right, int bottom, int mul){ doc_setmargins(handle, page, left, top, right, bottom, mul); });
+  vm.register_external("doc_setfont",         [this](int handle, int page, std::string_view font){ doc_setfont(handle, page, font); });
+  vm.register_external("doc_setlevel",        [this](int handle, std::string_view level){ doc_setlevel(handle, level); });
+  vm.register_external("doc_setlevelcoords",  [this](int handle, int left, int top, int right, int bottom){ doc_setlevelcoords(handle, left, top, right, bottom); });
+  vm.register_external("doc_show",            [this](int handle){ doc_show(handle); });
 
-  vm.registerExternalFunction("printdebug",          [this](Daedalus::DaedalusVM& vm){ printdebug(vm);           });
-  vm.registerExternalFunction("printdebugch",        [this](Daedalus::DaedalusVM& vm){ printdebugch(vm);         });
-  vm.registerExternalFunction("printdebuginst",      [this](Daedalus::DaedalusVM& vm){ printdebuginst(vm);       });
-  vm.registerExternalFunction("printdebuginstch",    [this](Daedalus::DaedalusVM& vm){ printdebuginstch(vm);     });
+  vm.register_external("exitgame",            [this](){ exitgame(); });
+
+  vm.register_external("printdebug",          [this](std::string_view msg){ printdebug(msg); });
+  vm.register_external("printdebugch",        [this](int ch, std::string_view msg){ printdebugch(ch, msg); });
+  vm.register_external("printdebuginst",      [this](std::string_view msg){ printdebuginst(msg); });
+  vm.register_external("printdebuginstch",    [this](int ch, std::string_view msg){ printdebuginstch(ch, msg); });
   }
 
-void Gothic::notImplementedRoutine(Daedalus::DaedalusVM& vm) {
+void Gothic::notImplementedRoutine(const std::string& fn) {
```

Comment:
string_view?

Reply:
That might be unsafe since I'd have to change the `std::set<string>` into a `std::set<string_view>` as well. The `string_view` passed to the missing external handler of phoenix' VM is a direct view onto the name of the underlying script symbol which is deallocated whenever the script goes out of scope (ie. a game reload).

I could re-write it to instead just keep a `std::set<uint32_t>` with the hashes of the function names if you'd like.

Reply:
Ah, I was very confused by the wording on cppreference. Turns out you have to specifically tell it to be a `std::set<std::string, std::less<>>`, otherwise the comparison operator is not "Transparent" which causes compilation to fail. It's implemented now.